packages/react-router/tests/link.test-d.tsx

```diff
@@ -216,6 +230,116 @@ test('when navigating from a route with no params to a route with params, params
   params.parameter(0).toEqualTypeOf<{}>()
 })
 
+test('when navigating from a route to a route with the same params, the params can be an object with optional properties', () => {
+  const TestLink = Link<RouteTree, '/invoices/$invoiceId', './edit'>
+  const params = expectTypeOf(TestLink).parameter(0).toHaveProperty('params')
+
+  params
+    .exclude<Function | boolean>()
+    .branded.toEqualTypeOf<{ invoiceId?: string | undefined } | undefined>()
+})
+
+test('when navigating from a route to a route with the same params, params is a function from params to optional next params', () => {
+  const TestLink = Link<RouteTree, '/invoices/$invoiceId', './edit'>
+  const params = expectTypeOf(TestLink)
+    .parameter(0)
+    .toHaveProperty('params')
+    .extract<Function>()
+
+  params.returns.branded.toEqualTypeOf<{ invoiceId?: string | undefined }>()
+  params.parameter(0).toEqualTypeOf<{ invoiceId: string }>()
+})
+
+test('when navigating to a union of routes, params is optional', () => {
```

Comment:
is this correct though? shouldn't `params` be `{ invoiceId: string } | `{ postId: string}`?

Reply:
yeah I was expecting this test:

https://github.com/TanStack/router/pull/1418/files#diff-921603b68301ae5e52534582aeb75f7defc2946a96ad3b268b28ad34ae5fec04R262

why would we need this test here? the test below already checks everything, right?

Reply:
I will double check but I believe this is slightly different as we're checking if it's optional. But it's a good point if this is intentional behaviour or not. Should it be optional for a union of different routes?

I find the unions a harder use case to think about if it should be optional or not. For example the default case `string` falls under the same code path and this makes sense to be optional

Reply:
`string` is basically considered to be any route and is therefore a union of everything